Add non-destructive baking for tables to ModelCommand

othercorey opened the pull request:
This is the code that affects the Table class generation. These are the only changes to the template and command code so far. 

This does not support importing existing properties. That can be added after this is reviewed.

othercorey committed:
Add non-destructive baking for tables to ModelCommand

othercorey commented on templates/bake/Model/table.twig:
```diff
@@ -14,13 +14,9 @@
-{% set uses = ['use Cake\\ORM\\Query;', 'use Cake\\ORM\\RulesChecker;', 'use Cake\\ORM\\Table;', 'use Cake\\Validation\\Validator;'] %}
-{{ uses|join('\n')|raw }}
+{% set generatedFunctions = ['initialize'] %}
+{% set classImports = fileBuilder.classImports(['Cake\\ORM\\Query', 'Cake\\ORM\\RulesChecker', 'Cake\\ORM\\Table', 'Cake\\Validation\\Validator']) %}
```
This always add the imports even if the functions aren't generated below. However, this is how it was done previously.

othercorey replied:
This is where I was talking about setting the imports in the command instead of the template.

othercorey commented on templates/bake/element/file_header.twig:
```diff
@@ -0,0 +1,5 @@
+declare(strict_types=1);
+
+namespace {{ namespace }};
+{{- Bake.concat("\n", [Bake.classUseStatements(imports.class ?? []), Bake.functionUseStatements(imports.function ?? []), Bake.constUseStatements(imports.const ?? [])], '\n\n') }}
```
The individual use statements could be rendered using an element instead of compiling with BakeHelper. 

The same goes for imported methods.

markstory replied:
I don't know if we need templates for use statements. There isn't much variance in how they are used.

othercorey replied:
The question isn't customization but what's the best way to maintain this. I wasn't sure if writing code gen into BakeHelper was right. You see I need to do some specific whitespace manipulation with it to support 0 statements.

ADmad commented on src/View/BakeView.php:
```diff
@@ -45,7 +45,8 @@ class BakeView extends TwigView
     {
         $this->setConfig('environment', [
-          'cache' => false,
+        'autoescape' => false,
```
Indentation looks off, surprising that phpcs didn't catch it.

othercorey replied:
These were not indented 4 spaces before for some reason which messed up my editor.

ADmad commented on templates/bake/Model/table.twig:
```diff
@@ -81,6 +77,7 @@ class {{ name }}Table extends Table
 {{- "\n" }}
 
 {%- if validation %}
+{% set generatedFunctions = generatedFunctions|merge(['validationDefalt']) %}
```
Typo in `validationDefalt`, should have been caught by a test case.

othercorey replied:
Will need many more tests once we decide on the template changes.

markstory commented on src/Command/ModelCommand.php:
```diff
@@ -1142,6 +1151,9 @@ public function buildOptionParser(ConsoleOptionParser $parser): ConsoleOptionPar
             'help' => 'The table name to use if you have non-conventional table names.',
+        ])->addOption('update', [
+            'boolean' => true,
+            'help' => 'Update existing file with generated code only.',
```
```suggestion
            'help' => 'Update generated methods in existing files. If the file doesn't exist it will be created.',
```

markstory commented on tests/TestCase/Command/ModelCommandTest.php:
```diff
@@ -1701,6 +1701,59 @@ public function testBakeWithRulesUnique()
+             * @param \Cake\ORM\RulesChecker $rules The rules object to be modified.
+             * @return \Cake\ORM\RulesChecker
+             */
+            public function buildRules(RulesChecker $rules): RulesChecker
```
It would be good to have coverage on custom methods like a finder. It would also be good to have coverage for overwriting existing methods like replacing `initialize()`.

othercorey committed:
Moed code generation into CodeGenHelper and simplified functions and parameters.

othercorey commented:
Moved code gen into separate helper and reduced the public API for generation code. Removed options like ignoring certain functions that can be added when something requires it.

Cleaned up template generation so feels less cluttered.

Added several tests for the various cases of overwriting and keeping different functions.

ADmad commented on src/View/Helper/CodeGenHelper.php:
```diff
@@ -0,0 +1,141 @@
+     * Builds php code from class, function and const imports.
+     *
+     * @param array<string|int, string> $classImports Class imports
+     * @param array<string|int, string> $functionImports Class imports
+     * @param array<string|int, string> $constImports Class imports
```
```suggestion
     * @param array<string|int, string> $functionImports Function imports
     * @param array<string|int, string> $constImports Constant imports
```

ADmad commented on src/View/Helper/CodeGenHelper.php:
```diff
@@ -0,0 +1,141 @@
+     * Builds array of php use statements from imports.
+     *
+     * @param string $prefix Prefix to put before import (function, const)
+     * @param array<string|int, string> $imports Class imports
```
```suggestion
     * @param array<string|int, string> $imports Imports
```
